visual studio for generated lib porblem

Hi serizba, thanks for your wounderful job.
It is ok to use cppflow2 to load the savedmodel and show the predication result from visual studio c++ generated **exe** file.
But, I met a problem after I turned the cpp file into dll and lib and made the implicit invocation of them. 
The console got stuck after showing  **Successfully opened dynamic library cudart64_101.dll**. Could you please help me to solve the problem? Thank you.

@ljn917 
Some of the errors:
 ```
'TFE_Context *cppflow::context::get_context(void)': symbol cannot be defined within namespace 'cppflow'
error C2888: 'TF_Status *cppflow::context::get_status(void)': symbol cannot be defined within namespace 'cppflow'
error C2888: 'cppflow::context::context(TFE_ContextOptions *)': symbol cannot be defined within namespace 'cppflow'
error C2065: 'get_status': undeclared identifier
error C2664: 'TFE_Context *TFE_NewContext(const TFE_ContextOptions *,TF_Status *)': cannot convert argument 2 from 'TF_Status *(__cdecl *)(void)' to 'TF_Status *'
There is no context in which this conversion is possible
```
Of course, I m gonna try the example.

@ljn917 
`E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(51,48): error C2888: 'TFE_Context *cppflow::context::get_context(void)': symbol cannot be defined within namespace 'cppflow'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(55,45): error C2888: 'TF_Status *cppflow::context::get_status(void)': symbol cannot be defined within namespace 'cppflow'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(60,55): error C2888: 'cppflow::context::context(TFE_ContextOptions *)': symbol cannot be defined within namespace 'cppflow'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(61,35): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(64,69): error C2664: 'TFE_Context *TFE_NewContext(const TFE_ContextOptions *,TF_Status *)': cannot convert argument 2 from 'TF_Status *(__cdecl *)(void)' to 'TF_Status *'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(64,60): message : There is no context in which this conversion is possible
1>D:\libtensorflow-gpu-windows-x86_64-2.3.0\include\tensorflow/c/eager/c_api.h(99,36): message : see declaration of 'TFE_NewContext'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(66,63): error C2664: 'TFE_Context *TFE_NewContext(const TFE_ContextOptions *,TF_Status *)': cannot convert argument 2 from 'TF_Status *(__cdecl *)(void)' to 'TF_Status *'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(66,54): message : There is no context in which this conversion is possible
1>D:\libtensorflow-gpu-windows-x86_64-2.3.0\include\tensorflow/c/eager/c_api.h(99,36): message : see declaration of 'TFE_NewContext'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(68,31): error C2664: 'bool cppflow::status_check(TF_Status *)': cannot convert argument 1 from 'TF_Status *(__cdecl *)(void)' to 'TF_Status *'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(68,22): message : There is no context in which this conversion is possible
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(17,17): message : see declaration of 'cppflow::status_check'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(71,53): error C2888: 'cppflow::context::context(cppflow::context &&) noexcept': symbol cannot be defined within namespace 'cppflow'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(76,64): error C2888: 'cppflow::context &cppflow::context::operator =(cppflow::context &&) noexcept': symbol cannot be defined within namespace 'cppflow'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\context.h(81,32): error C2888: 'cppflow::context::~context(void)': symbol cannot be defined within namespace 'cppflow'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(133,83): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(135,81): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(135,118): error C2679: binary '=': no operator found which takes a right-hand operand of type 'initializer list' (or there is no acceptable conversion)
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\include\memory(1183,17): message : could be 'std::shared_ptr<TFE_TensorHandle> &std::shared_ptr<TFE_TensorHandle>::operator =(std::shared_ptr<TFE_TensorHandle> &&) noexcept'
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\include\memory(1172,17): message : or       'std::shared_ptr<TFE_TensorHandle> &std::shared_ptr<TFE_TensorHandle>::operator =(const std::shared_ptr<TFE_TensorHandle> &) noexcept'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(135,118): message : while trying to match the argument list '(std::shared_ptr<TFE_TensorHandle>, initializer list)'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(136,31): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(167,83): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(168,31): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(181,81): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(181,118): error C2679: binary '=': no operator found which takes a right-hand operand of type 'initializer list' (or there is no acceptable conversion)
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(181,118): message : while trying to match the argument list '(std::shared_ptr<TFE_TensorHandle>, initializer list)'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(182,31): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(186,38): error C2065: 'get_context': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(186,71): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(187,31): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(189,26): error C3536: 'op': cannot be used before it is initialized
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(189,61): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(190,31): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(195,43): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(196,31): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(197,24): error C2664: 'void TFE_DeleteOp(TFE_Op *)': cannot convert argument 1 from 'int' to 'TFE_Op *'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(197,22): message : Conversion from integral type to pointer type requires reinterpret_cast, C-style cast or function-style cast
1>D:\libtensorflow-gpu-windows-x86_64-2.3.0\include\tensorflow/c/eager/c_api.h(252,28): message : see declaration of 'TFE_DeleteOp'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(205,86): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(207,91): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(209,31): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(237,77): error C2065: 'get_status': undeclared identifier
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(237,107): error C2679: binary '=': no operator found which takes a right-hand operand of type 'initializer list' (or there is no acceptable conversion)
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(237,107): message : while trying to match the argument list '(std::shared_ptr<TF_Tensor>, initializer list)'
1>E:\cppflow-cppflow2\cppflow-cppflow2\include\cppflow\tensor.h(238,35): error C2065: 'get_status': undeclared identifier`

--- a/include/cppflow/context.h
+++ b/include/cppflow/context.h
@@ -41,14 +41,20 @@ namespace cppflow {
     };
 
     // TODO: create ContextManager class if needed
-    inline context global_context;
+    // Set new context, thread unsafe, must be called at the beginning.
+    //  TFE_ContextOptions* tfe_opts = ...
+    //  cppflow::get_global_context() = cppflow::context(tfe_opts);
+    inline context& get_global_context() {
+        static context global_context;
+        return global_context;
+    }
 
 }
 
 namespace cppflow {
 
     inline TFE_Context* context::get_context() {
-        return global_context.tfe_context;
+        return get_global_context().tfe_context;
     }
 
     inline TF_Status* context::get_status() {

--- a/include/cppflow/context.h
+++ b/include/cppflow/context.h
@@ -65,8 +65,8 @@ namespace cppflow {
     inline context::context(TFE_ContextOptions* opts) {
         auto tf_status = context::get_status();
         if(opts == nullptr) {
-            std::unique_ptr<TFE_ContextOptions, decltype(&TFE_DeleteContextOptions)> opts(TFE_NewContextOptions(), &TFE_DeleteContextOptions);
-            this->tfe_context = TFE_NewContext(opts.get(), tf_status);
+            std::unique_ptr<TFE_ContextOptions, decltype(&TFE_DeleteContextOptions)> new_opts(TFE_NewContextOptions(), &TFE_DeleteContextOptions);
+            this->tfe_context = TFE_NewContext(new_opts.get(), tf_status);
         } else {
             this->tfe_context = TFE_NewContext(opts, tf_status);
         }

